Add more details about example-cnf behavior in README and update diagram

ramperher opened the pull request:
- Including in README what are the pods deployed by each operator, briefly describing them.
- Updating the diagram to include the pod distribution on each module defined
- Updating the diagram to correct the SRIOV network distribution in the scenario, according to what it's observed in example-cnf jobs

ramperher commented:
@tkrishtop @capolrik , I'd like that you take a look to this and review it, just to see if I am correct with what I've been analyzing. And also, to confirm the SRIOV network distribution in the scenario. According to the current diagram we have, it looks like all components are using both SRIOV networks, but it's not true, just take a look to the network definition in one of the example-cnf deployments I've seen today (and I've verified it's the same in others):

TRex pod is just connected to intel-numa0-net2 network in both interfaces

```
$ oc describe pod trexconfig-6b747d779-ld7fw | less
                    },{
                        "name": "example-cnf/intel-numa0-net2",
                        "interface": "net1",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:0a.2"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net2",
                        "interface": "net2",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:0a.7"
                            }
                        }
                    }]
```

TestPMD LB has four interfaces, two connected to net1, and other two to net2:

```
$ oc describe pods loadbalancer-5c47c445f7-vf6b9 | less
                    },{
                        "name": "example-cnf/intel-numa0-net2",
                        "interface": "net1",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:0a.0"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net2",
                        "interface": "net2",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:0b.2"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net3",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:03.5"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net4",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:03.0"
                            }
                        }
                    }]
```

And finally, CNF Apps are just connected to intel-numa0-net1:

```
$ oc describe pod testpmd-app-646d7bb697-6f7m5 | less
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net1",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:03.4"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net2",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:03.2"
                            }
                        }
                    }]

$ oc describe pod testpmd-app-646d7bb697-zgsdb | less 
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net1",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:03.7"
                            }
                        }
                    },{
                        "name": "example-cnf/intel-numa0-net1",
                        "interface": "net2",
                        "dns": {},
                        "device-info": {
                            "type": "pci",
                            "version": "1.0.0",
                            "pci": {
                                "pci-address": "0000:37:02.4"
                            }
                        }
                    }]
```

So, my assumption is that the network setup looks like to what I'm drafting in this change, rather than what we currently have.

Don't know what you think, of course comments are welcome to clarify and improve this. That's the main goal!

capolrik commented:
To be honest, I though there was one SRIOV network for the RX packet and one for the TX so I am little confused.
Anyway, that's a good thing to start digging into the details of the SR-IOV networks. It will help us understand why we got so high packet loss (it should be less than 1% with this bandwith)

tkrishtop reviewed (changes requested):
Hi @ramperher I believe the diagram modification is wrong:

- The pod renaming is wrong. The main flow is trex pod sending the traffic to testpmd pod, not to testpmd-lb pod. @capolrik knows this flow really in details.
- The network changes look weird. My diagram [copies the presentation of Saravanan](https://docs.google.com/presentation/d/17W-kbA3oxsq1O8AJhXEyOM6vKMvQCd27l3hHptpOxKs/edit#slide=id.g9ceb35f478_0_34), I just added the arrows with the traffic direction. Why do think the Saravanan's flow presentation is wrong?

ramperher commented:
> Hi @ramperher I believe the diagram modification is wrong:
> 
> * The pod renaming seems to be not correct. The main flow is the trex pod sending the traffic to testpmd pod, not to testpmd-lb pod. @capolrik knows this flow really in detail.
> * The network changes look weird. My diagram [copies the presentation of Saravanan](https://docs.google.com/presentation/d/17W-kbA3oxsq1O8AJhXEyOM6vKMvQCd27l3hHptpOxKs/edit#slide=id.g9ceb35f478_0_34), I just added the arrows with the traffic direction. I would think that the flow documented by Saravanan is correct.
> 
> To get a valid visual representation of the network connections, we could deploy [network observability operator](https://docs.openshift.com/container-platform/4.13/network_observability/understanding-network-observability-operator.html) in BOS2, and check the real flow directly on the OCP console. The operator is easy to deploy, I tested it in the past.

I also supposed that Saravanan's workflow should be correct, but the network interfaces of the components we're deploying are not saying the same. If that's correct, then we have two options: 1) work in the mode we're deploying right now, or 2) change the SRIOV network config to match Saravanan's scenario.

Network observability is a good option. I'll move this to draft while we figure out this.

ramperher commented:
After finding some interesting behavior in https://issues.redhat.com/browse/CILAB-1489, I've updated the README with my findings and I've moved this to open, so that it can be re-reviewed again, but @tkrishtop, @capolrik , I'd like to discuss with you about this just to check that everything is clear and to confirm if something needs to be updated, because again, I've confirmed that the SRIOV setup we have is not the same than the one depicted by Saravanan, and I don't know if we want this or not. I'll probably ask Franck in the next sync about this to check his opinion. Thanks.

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
 
-* testpmd-operator
 * trex-operator
+    * It provides TRex Traffic Generator, decomposed in two components: a TRex server, deployed as `trexconfig-<x>` pod, which takes care of deploying and configuring a TRex server instance, and a TRex application, deplayed as `trex-app` job, that starts TRex server, generating traffic towards the system under test.
```
it could be nice to speak about TRex CR, pinning the config, and TRex operator, ensuring reconciliation via controller-manager pod by running short-living Trex jobs.

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
 * trex-operator
+    * It provides TRex Traffic Generator, decomposed in two components: a TRex server, deployed as `trexconfig-<x>` pod, which takes care of deploying and configuring a TRex server instance, and a TRex application, deplayed as `trex-app` job, that starts TRex server, generating traffic towards the system under test.
+    * In `trexconfig-<x>` logs, you can see the trex statistics printed periodically. The summary of the test execution can be seen at the end of the `trex-app` job logs.
+    * In our case, as we will see later on, the `trexconfig-<x>` pod has two interfaces connected to the same SRIOV network.
```
we could keep docs a bit more factual by omiting "In our case, as we will see later on,"

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
+    * In `trexconfig-<x>` logs, you can see the trex statistics printed periodically. The summary of the test execution can be seen at the end of the `trex-app` job logs.
+    * In our case, as we will see later on, the `trexconfig-<x>` pod has two interfaces connected to the same SRIOV network.
 * testpmd-lb-operator
+    * Its main component is a modified TestPMD instance, implementing a custom load balancing forwarding module, which is eventually used to perform load balancing between the ports of the deployed pod, called `loadbalancer-<x>`.
```
"perform load balancing between the ports of the deployed pod" - do we speak about cnf-app pod here or about another pod?

ramperher replied:
we talk about loadbalancer pod, but I'll try to rewrite this to make it clearer

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
 * testpmd-lb-operator
+    * Its main component is a modified TestPMD instance, implementing a custom load balancing forwarding module, which is eventually used to perform load balancing between the ports of the deployed pod, called `loadbalancer-<x>`.
+    * This pod is composed by two containers: `loadbalancer`, which performs the load balancing forwarding, and `listener`, an auxiliary module that is listening to the CNFAppMac component created by the cnf-app-mac-operator to retrieve the MAC addresses of the TestPMD instances launched by the testpmd-operator (i.e. the CNF Applications), then serving this information to the `loadbalancer` conta ...
+    * To see the TestPMD statistics printed periodically for this module, you can rely on `loadbalancer` container logs.
```
"you can rely on `loadbalancer` container logs" - we could probably name logfile here. Since we have both loadbalancer and loadbalancer-lb, it's not really clear what exactly to look for

ramperher replied:
I'll rewrite this, but the only log related to LB, apart from controller-manager, is loadbalancer. Loadbalancer-lb does not exist

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
+    * The `loadbalancer-<x>` pod is not currently replicated; only one instance is deployed. It's supposed that, up to now, it's not belonging to any update-upgrade process of the example-cnf, so that it has to be deployed in an isolated worker node, whereas the other components of this CNF (TRex and CNF Application) have to be deployed in a different worker node.
+* testpmd-operator
+    * Final application, also known as CNF Application, which is a standard TestPMD instance using the default MAC forwarding module. Two replica pods are deployed, called `testpmd-app-<x>`, having each of them, in our case, two ports connected to the same SRIOV network, but different than the SRIOV network used by TRex.
+    * To see the TestPMD statistics printed periodically for this module, you can rely on `testpmd-app-<x>` container logs. Each log will offer you the statistics of each replica pod.
```
When discussing an operator, we could probably discuss all the components, not just a pod. CR, pinning the state, is really the key here. Also, we deploy both loadbalancer and loadbalancer-lb, and you're discussing only loadbalancer-lb speaking about loadbalancer pod. That makes things pretty unclear

ramperher replied:
honestly, I don't get your point about loadbalancer and loadbalancer-lb :sweat_smile:  there's just one loadbalancer and it's called loadbalancer. Maybe you're referring to TestPMD and TestPMD LB, I'll try to refactor that to make it clearer

tkrishtop commented on README.md:
```diff
@@ -3,16 +3,27 @@ Example CNF
+    * Final application, also known as CNF Application, which is a standard TestPMD instance using the default MAC forwarding module. Two replica pods are deployed, called `testpmd-app-<x>`, having each of them, in our case, two ports connected to the same SRIOV network, but different than the SRIOV network used by TRex.
+    * To see the TestPMD statistics printed periodically for this module, you can rely on `testpmd-app-<x>` container logs. Each log will offer you the statistics of each replica pod.
 * cnf-app-mac-operator
+    * Auxiliary operator used to deploy a resource called CNFAppMac, which is a wrapper created for each `testpmd-app-<x>` and linked to them, and that are used to extract the network information of these pods (MAC and PCI addresses), to be offered to other components of the solution, such as the TestPMD Load Balancer.
```
the same - cnf-app is an operator, not just a pod

tkrishtop commented on README.md:
```diff
@@ -104,12 +115,234 @@ Ansible based automation
+- Load balancer mode (`enable_lb: true`):
+    - Represents the behavior described above, with a TestPMD instance acting as load balancer between TRex and the CNF Applications.
+- Direct mode (`enable_lb: false`):
+    - The TestPMD instance is not used, then testpmd-lb-operator is not deployed.
```
"The TestPMD instance is not used, then testpmd-lb-operator is not deployed" - do we speak here about loadbalancer or about loadbalancer-lb? Right above, both are mixed up

ramperher replied:
I made a mistake here, I will fix that

tkrishtop commented on README.md:
```diff
@@ -104,12 +115,234 @@ Ansible based automation
+    - The `loadbalancer-<x>` pod is deployed in a different worker than `trexconfig-<x>` and `testpmd-app-<x>`.
+    - There are no constraints for the location of `trexconfig-<x>` and `testpmd-app-<x>` pods, so they may live in the same worker.
+- Direct mode:
+    - Since load balancer is not used, in this scenario, `trexconfig-<x>` and `testpmd-app-<x>` pods are placed in different worker nodes.
```
Right above we only discussed pods, and now jump into config which is pinned by CR but was not even mentionned before. We could probably add more details for each operator, adding some pictures or more detailed explanations. 

![image](https://github.com/openshift-kni/example-cnf/assets/26544656/eac7fd3f-1647-4de3-90d0-56339bc84f9f)

ramperher replied:
I've included the consideration for CRs, I'll add also some diagrams. Regarding affinity rules, we need to talk about pods in that point since these are specs for the pods, I'll keep using this for the moment

ramperher replied:
Since the diagram is in fact common for all operators, since it depicts the reconciliation loop, I'll add it to the section where we talk about operator-sdk

tkrishtop reviewed (commented):
Hi Ramon, I have reviewed the first part of the PR, and will check the second part this afternoon. 

My main concern so far is that we only speak about pods while the actual structure is operator, with the main logic inside the CR and reconciliation (which triggers job sending traffic for the TRex, for example).

ramperher commented:
@tkrishtop @capolrik I've updated the PR according to your comments and also taking into account the new SRIOV setup, feel free to review it again, thanks!

capolrik commented on README.md:
```diff
@@ -3,23 +3,47 @@ Example CNF
+        * In `testpmd-lb-operator-controller-manager-<x>` pod, you can see the execution of the Ansible playbooks that ensures the reconciliation loop of the operator.
+    * The LoadBalancer is not currently replicated; only one instance is deployed. It's supposed that, up to now, it's not belonging to any update-upgrade process of the example-cnf, so that it has to be deployed in an isolated worker node, whereas the other components of this CNF (TRex and CNF Application) have to be deployed in a different worker node. Pod anti-affinity rules ensure this setup.
+* testpmd-operator
+    * Final application, also known as CNF Application, which is a standard TestPMD instance using the default MAC forwarding module. It uses two components:
```
'the default MAC forwarding module'
Do you mean 'by default' or the regular module?

capolrik reviewed (approved):
Very good!

tkrishtop reviewed (approved):
Hi @ramperher really great change, thank you for all the hard work investigating the traffic flow! 

As we discussed, we could add some improvement in the follow-up changes:

- operator image could be customized for every operator
- it would be nice to add a diagram for the affinity rules
- it would be nice to separate traffic flow diagrams for with and without `enable_lb`.

Thank you again for all the work, really precious change and great investigation work, let's merge it asap! :rocket:

ramperher commented:
> Hi @ramperher really great change, thank you for all the hard work investigating the traffic flow!
> 
> As we discussed, we could add some improvement in the follow-up changes:
> 
> * operator image could be customized for every operator
> * it would be nice to add a diagram for the affinity rules
> * it would be nice to separate traffic flow diagrams for with and without `enable_lb`.
> 
> Thank you again for all the work, really precious change and great investigation work, let's merge it asap! 🚀

Sure, we can keep improving the docs with all these remaining points to check, I'll note them to include them in the short term. Thanks for reviewing!
